### DISCUSSION TITLE
Dungeon: Improve dungeon speed with flash (#170)

### DISCUSSION THREAD

**Discussion by Farigh** [Farigh/pokeclicker-automation#173] [PR discussions]
> So, for conditions, what players might want is:
> 
>   - Do everything (all fights, then all chests), then go to the boss
>   - Do all fights, then go to the boss (ignoring chests)
>   - Do all chests, then go to the boss (ignoring fights)
>   - Go directly to the boss (ignoring fights and chests)
> 
> 
> Special:
>   - Do only the chests (ignoring fights and the boss), to farm Darumaka Zen
>   - Do only the fights (ignoring chests and the boss), to capture a non-boss pokemon (this is probably triggered from another function, like a future "Capture all pokemons" focus)
>   - Do all fights, then all chests, then reset. Like the previous one, but even more niche.
>   - Do nothing (ignore all). This one is the only trully useless.
> 
> 
> By the pidgeonhole principle, if we want to allow 5 posibilities, two switches are not enough. We need either a third switch or a dropdown.
> 
> I suggest to have three switches:
> 
>   - Do all fights
>   - Collect all chests
>   - Kill boss
> 
> 
> However, this means changing the previous switches meanings. If all switches are off, the dungeon should not be startable (there is nothing to do anyway)

**Comment by Farigh** [comment]
> I suggest to have three switches:
> 
>   - Do all fights
>   - Collect all chests
>   - Kill boss
> 
> However, this means changing the previous switches meanings. If all switches are off, the dungeon should not be startable (there is nothing to do anyway)

I think we only need an additional switch :
  - Boss rush
  - Skip chest
  - Skip unnecessary fights (the new one)

**Reply by Daniferrito** [reply]
What combination would allow to get only chests, and ignore the boss? (to farm Darmanitan Zen and Galarian Darmanitan Zen)

Isn't skip unnecessary fights just the same as Boss rush? I do not understaund what it would do

**Reply by Farigh** [reply]
Hum you're right it would be equivalent in this case ...

I feel like I did not understand your kill-boss option then.
Killing the boss is the exit point of the dungeon, so I don't see what it would condition ? Do you mean somthing like `Exit the dungeon without killing the boss` or something ?

**Reply by Daniferrito** [reply]
> Exit the dungeon without killing the boss

Exactly. After gathering 10 to 16 chests, the boss might take a LOT of time to kill, with almost no benefit. Just leaving the dungeon and reentering is faster, and once the dungeon is cleared 500 times, extra clears do nothing (archievement is already gotten, the loot changes are already maxed) (it does make the dungeon smaller at 1000 clears, but that is actually worse for farming Darumaka Zen, as it means less chests per entry)

**Reply by Farigh** [reply]
Only adding a new button should still be enough:
  - Boss rush
  - Skip chest
  - Exit the dungeon without killing the boss (the new one)

**Reply by Daniferrito** [reply]
Ok, lets see what each combination would do (mostly for myself):
Each combination is the three toggles in order (boss rush, skip chest, dont kill boss), expressed as 1 (toggle on) or 0 (toggle off)
* 000 - Kill everything, collect all chests, kill boss
* 001 - Kill everything, collect all chests, exit dungeon
* 010 - Kill everything, ignore chests, kill boss
* 011 - Kill everything, ignore chests, exit dungeon
* 100 - Ignore fights, collect all chests, kill boss (currently this one will collect whatever chests it has seen and ignore hidden ones)
* 101 - Ignore fights, collect all chests, exit dungeon
* 110 - Ignore fights, ignore chests, kill boss
* 111 - Ignore fights, ignore chests, exit dungeon (this one should disable the option to start the dungeon, as it will exit inmediatly anyways)

At about the third line, I realized that this is the three switches I suggested, but inverted. Works the same way, but keeps the meaning of the two already existing switches.
I suggest to make a note that "boss rush" means to ignore as many fights as possible, either in the name itself, or in the tooltip.

**Reply by Daniferrito** [reply]
With the "dont kill boss" option off (as it should be by default), the only change from current behaviour would be to collect all chests if boss rush is on (and probably only for dungeons with flash unlocked, as base dungeons cant be optimized better)

Probably do check for more visible chests right before the boss kill, as some might be accessible without a fight, and opening half of them does discover everything.

**Reply by Farigh** [reply]
It might be a good idea to rename the boss rush button instead of only rewording the tooltip.
Maybe `Fight every battles` to stick with your initial suggestion ?

⚠️ Internal settings should be added to allow the Focus.Quest and Focus.Achievement to bypass the `Skip boss` setting

**Reply by Daniferrito** [reply]
Both of this suggestions were handled, the button got renamed to `Skip as many fights as possible`, and the internal settings you handled yourself

**Comment by Farigh** [comment]
@Daniferrito you've not answered my last review comments, nor have I been requested a review either. 
Are you done with this feature ?

If so, can you resolve my addressed comments and reply to the ones you didn't

**Reply by Daniferrito** [reply]
I was also waiting, so thanks for pointing it out.

I will make sure to review all sugestions and make sure they are addressed, and then request a review.

**Reply by Farigh** [reply]
@Daniferrito I unresolved 2 comments that were not addressed, could you look at them if needed, or give an explanation for not changing those ?

**Reply by Daniferrito** [reply]
@Farigh In case you didnt see them, I commented on both.
In short, I believe changes to be unnecessary, but can add them if you insist (or maybe I'm missunderstaunding something)

**Reply by Farigh** [reply]
I don't see those :/
Did you, by any chance, start a review and did not summit it ?

**Reply by Daniferrito** [reply]
No, they are in the PR, and I answered there. PR #170 
I dont see any way to link to my answers, but I will copy it here:

First one you added a suggestion about my comment in line 340 (or 342, changes depending on where you are looking):
> `// Do not call DungeonRunner.nextFloor() directly, as it has no checks`

Your comment:
> That was the point, if we are here, we should always be in the state that would lead directly to this call (and avoid all those check overhead).
If it's not the case, we should do this check manually, since resetting boss chests and boss position would make us in an undefined state

My answer (with more comments in the middle):
> I dont understaund what the remaining issue here is. Do you mean that if foundFloorEndTile has a value, then by definition it must be accessible (since we check for accesibility when adding it), so we can call nextFloor without going through handleClick()?
I kind of agree, but I still prefer to keep the handleClick there, as it helps detect future problems if someone edits this code further. It is a hard problem to detect visually, as it moves so fast, and simply warping to the next floor hides the probem too well.

Second suggestion is method `__internal__ensureTheBoardIsInAnAcceptableState()`:
```
if (DungeonRunner.map.hasAccessToTile(currentLocation))
{
    this.__internal__markCell({ ...currentLocation, tile });
}
```

Your comment:
> any reason to do it only if the player has access to it before marking it, in this case ?
The other calls are not protected this way

My answer:
>I dont understaund the unresolve here.

>This check needs to be here, or the automation will think that tiles discovered through opening chests are accessible (this can only happen if the user interacts manually, but this is in the recovery method, it gets called when we suspect there was manual intervention). Without the check, it will "see" that the ladder/boss is visible, and try to walk to it, but the game wont let us. This will cause an infinite loop (until the timer runs out and we are forced to restart)

>The other two places this check is not done because it is not necessary. On those cases, we are either on top of the cell (no flash), so isVisited will be true, or we are touching the cell (with flash), so one of the surounding cells is visited. Those are literaly the two conditions that hasAccessToTile checks. I can add the check there, but it is superfluous.

>If you want me to add the checks to the other cases, let me know and I will add them. If you agree that they are not needed, just resolve this conversation. If the issue is something else and I'm missunderstaunding, just let me know.

In the PR, the suggestions are shown kinda spread out, but seem to be condensed at the bottom. However, they are easy to see in the "Files changed" view, as those two are the only one still open.

**Reply by Farigh** [reply]
I don't know why they do not appear on my side, if you did summit those then ...

I still don't see them, neither on the main page, nor in the files changed section : 
![image](https://user-images.githubusercontent.com/11090416/199555400-215b2f63-e281-4096-91d8-c71ce02c5370.png)

Are you sure they are not `Pending` ?

**Reply by Farigh** [reply]
For the 1st case I did it for performance concerns, as I run this on my rpi4 and the game is really poorly optimised ....
But it can stay this way for now, since this code is not called that often and should have a negligible impact.

For the 2nd case you are right, I should have payed more attention

**Reply by Daniferrito** [reply]
About the screenshot, you may need to scroll down more. I also see parts of the threads (but only my parts), completely split, but scrolling down more I can see the complete thread.

Or maybe because you did not interact with them recently, they are still showing as resolved to you so they are still collapsed?

In any case, it seems like we are seeing different things.

<details>
  <summary>My view (in spoiler because the screenshot is very tall)</summary>

![screencapture-github-Farigh-pokeclicker-automation-pull-170-2022-11-02-18_27_29](https://user-images.githubusercontent.com/2537742/199559902-4827c521-ce27-4abe-b83b-9ae328f3b98c.png)
</details>

> Are you sure they are not Pending ?

I see no status and no way of changing status.

> For the 1st case I did it for performance concerns

It should run once per dungeon with multiple floors, and because currently those cannot have flash, it will exit on the first check (player has already visited it)

> For the 2nd case you are right, I should have payed more attention

No problem

> I run this on my rpi4 and the game is really poorly optimised

I did not think about running it on my rpi, but this is a good candidate for it. I agree on the poor optimization, as it usually takes quite a chunk of my 9900k.

**Reply by Farigh** [reply]
From your screenshot, they are indeed pending : 
![image](https://user-images.githubusercontent.com/11090416/199564266-8b994c89-84d3-4683-aafd-1a2537caa73a.png)

It means you clicked : 
![image](https://user-images.githubusercontent.com/11090416/199564453-1fab0232-19dd-4a12-99fb-7b2e47336080.png)

instead of : 
![image](https://user-images.githubusercontent.com/11090416/199564486-b7bc3c6e-1a39-42ad-ae16-8b8f0512a474.png)


You can submit your pending comments from the `Files changed` by clicking `Review changes`, and then the `Submit review` button (leave `Comment` checked in this case) : 
![image](https://user-images.githubusercontent.com/11090416/199564793-be2cfa11-dfa9-4a4c-b7de-cc1cbc77f394.png)

**Reply by Daniferrito** [reply]
It seems like just posting a top-level review comment marked reopened the suggestions. I'm still not clear how Github is handling what to show to which people.

Can you confirm that you see the suggestion threads on your side now?

**Reply by Farigh** [reply]
Yup that's alright now.

GitHub has 2 comment types: 
- Single comments, that are posted right away and are the default when you reply from the front-page of a PR (in the `reply` field) :
  ![image](https://user-images.githubusercontent.com/11090416/199569927-098e468f-b35b-437b-a8ad-d4b2d1b219ec.png)
  With a single button to send the reply
- Review comments, from the `Files changed` tab, where you can either, that presents the 2 buttons I mentioned in my previous comment, to either:
  - Start a review and stack-up as many comments as needed, delete/edit/amend them and post the whole review once you're done (and won't be displayed to anyone but you until then)
  - Post single comments that are sent immediately, like in the first case

**Reply by Daniferrito** [reply]
> Single comments, that are posted right away and are the default when you reply from the front-page of a PR (in the reply field) :

I did reply from the main page of the PR, not the review side. I will make sure in the future that the Pending is not there, and to comment from the files changed side if it is there. I simply thought the pending meant that you (as the one that made the original comment) were suposed mark that suggestion as done.

**Reply by Farigh** [reply]
That's wierd ... maybe you had a review opened that might have explained this behaviour....

At least now you're aware of this 😅

I'll merge the changes, thanks again for your contribution

**Comment by Farigh** [comment]
Locking this discussion as the PR was merged
